src/main/java/com/griddynamics/cloud/learning/service/UserService.java

```diff
@@ -0,0 +1,80 @@
+package com.griddynamics.cloud.learning.service;
+
+import com.griddynamics.cloud.learning.dao.Permission;
+import com.griddynamics.cloud.learning.dao.domain.Role;
+import com.griddynamics.cloud.learning.dao.domain.User;
+import com.griddynamics.cloud.learning.dao.repository.RoleRepository;
+import com.griddynamics.cloud.learning.dao.repository.UserRepository;
+import com.griddynamics.cloud.learning.web.dto.UserDto;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+@Service
+public class UserService implements UserDetailsService {
+
+    private UserRepository userRepository;
+
+    private RoleRepository roleRepository;
+
+    private PasswordEncoder encoder;
+
+    public UserService(UserRepository userRepository, RoleRepository roleRepository, PasswordEncoder encoder) {
+        this.userRepository = userRepository;
+        this.roleRepository = roleRepository;
+        this.encoder = encoder;
+    }
+
+    @Override
+    @Transactional
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = this.userRepository.findUserByUsername(username).orElseThrow(
+                () -> new UsernameNotFoundException("No user found with username " + username));
+
+        Set<GrantedAuthority> authorities = new HashSet<>();
+        for (Permission permission : user.getRole().getPermissions()) {
+            authorities.add(new SimpleGrantedAuthority(permission.toString()));
+        }
+
+        return new org.springframework.security.core.userdetails.User(user.getUsername(), user.getPassword(), authorities);
+    }
+
+    @Transactional
+    public User save(UserDto user) {
+
+        Optional<Role> existingRole = roleRepository.findAllBy().stream()
```

Comment:
Why do we need to filter for nonNull objects?
What does a second filter?

Btw, I think it will be easier to explicitly override findAll() method in RoleRepository to return List instead of Iterable so that you can create stream directly